### ISSUE TITLE
Some notes on using oxigraph in the browser/WebWorker

### ISSUE THREAD

**Issue by bollwyvl** [oxigraph/oxigraph#128]
Thanks for oxigraph!

I have some (unfortunately not yet public) code using `oxigraph.js` in a [JupyterLab](https://github.com/jupyterlab/jupyterlab)/[JupyterLite](https://github.com/jupyterlite/jupyterlite) extension and it works great! I've suffered through a number of in-browser graph stores, and am blown away at how robustly this one seems to work. The only thing I'm really missing (maybe just haven't found it) is prepared queries/bindings, but _I can deal_, given all the other goodness oxigraph provides!

Anyhow, an overview of my setup:

- running in the browser
- bundling with WebPack 5: I don't have a lot of choice in the matter, but has come in handy, because:
  - to get it working, I am using [string-replace-loader](https://github.com/Va1/string-replace-loader) on `oxigraph.js` to:
    - patch _out_ 
      - the weird `String.raw` dependency
      - `require('fs')` and and manually load in the wasm `bytes` with [arraybuffer-loader](https://github.com/pine/arraybuffer-loader)
    - patch _in_ 
      - the browser's native `Text(De|En)coder`s instead of `util`
- wrapping oxigraph in a webworker with [comlink](https://github.com/GoogleChromeLabs/comlink)
  - ran into some (unrelated) issues here, so had to do some hacks :blush:
- adding JSON-LD (#45) via [https://github.com/digitalbazaar/jsonld.js/](https://github.com/digitalbazaar/jsonld.js/) The Wrong Way by (expensively) marhalling through nquads
- adding some more typescript types (e.g. on `.query`) as the pointers to make it across the WebWorker barrier

I'm wondering if there is interest, aside from LD, as that's probably going to be better in rust, in:
- making a standalone package for others
- PR-ing an async worker implementation here on this repo

Thanks again!

**Comment by Tpt** [comment]
Hi! Thank you for your interest in Oxigraph.

> I have some (unfortunately not yet public) code using oxigraph.js in a JupyterLab/JupyterLite extension and it works great! 

It's an amazing project! I am eager to try it when it's released!

> I've suffered through a number of in-browser graph stores, and am blown away at how robustly this one seems to work. 

Thank you! Rust is a great tool for that!

> The only thing I'm really missing (maybe just haven't found it) is prepared queries/bindings, but I can deal, given all the other goodness oxigraph provides!

Yes, they are not implemented yet (but it is definitely something that should be done). Oxigraph JS is currently in a kind of "proof of concept" state (I never actually used it in a real project), so I guess there are a lot of other rooms for speed improvements. The current JS <-> Rust conversions are quite heavy and slow. Help would be very much welcome in this area. [If you want to have a look at this topic, this research paper is quite interesting](https://openreview.net/pdf?id=CXLmXMb2TJ).

> I'm wondering if there is interest, aside from LD, as that's probably going to be better in rust, in:
> making a standalone package for others
>  PR-ing an async worker implementation here on this repo

Both would be amazing. Thank you! [The build script](https://github.com/oxigraph/oxigraph/blob/ec2b3072555a2449d0bc9951580344f3693be005/js/package.json#L12) is currently targetting NodeJS only, there is maybe a way to build an `oxigraph` package that would work on both NodeJS and Web. I never took time to explore it. [Here is the build system documentation](https://rustwasm.github.io/docs/wasm-pack/). About the async worker implementation, if it's web only it might be nice to have it into a stand-alone package wrapping the basic one. A PR is this main repo is very welcome except if you prefer to develop it externaly (it's what I did [for  the python rdflib wrapper](https://github.com/oxigraph/oxrdflib)).

Thank you again.

**Comment by bollwyvl** [comment]
Thanks for the speedy reply.

> when it's released

The current effort has as a goal enabling [jupyterlab-knowledge-graph](https://github.com/agoose77/jupyterlab-knowledge-graph/issues/7). In a classic client-server setting (JupyterLab), `pyoxigraph` would provide a better out-of-the-box-plus-one (over `rdflib-sqlalchemy` on `sqlite3`) for a "real" jupyter server until sharing data between many users, megatriples, etc. are required and a distributed SPARQL store will make sense.

But the JupyterLite part is the tricky/fun part!

> for the python rdflib wrapper

That's a great example: yet again, I've suffered through a number of graph store integrations on the python side that ignore the existing decade of work by rdflib, so thank you for making things that would Just Work (Better) with existing code. 

As a heads-up: I will probably package both the rdflib wrapper and the underlying python bindings for [`conda-forge`](https://conda-forge.org/) when i get a chance. I anticipate packaging it being much easier than some of the other stores!

>  not much familiar with WebWorkers

Right, we can't all know all the languages and runtimes! 

The raw WebWorker API isn't much fun... I've taken to using `comlink`, such as for [this `sqlite.js` wrapper](https://github.com/nrbgt/janki/blob/gh-6-add-cards/js/jupyterlab-sqlite3/src/sqlite.worker.ts) from which I cribbed almost all the boilerplate. But it's rather a heavy opinion, so indeed might make the most sense for me to split up my prototype into several packages, one of which is an opinionated `oxigraph` wrapper. Once we can talk about it more directly, we can decide what can be upstreamed.

> current JS <-> Rust conversions are quite heavy and slow. Help would be very much welcome 

Ha, to the point above: I am very much a rust dilettante, knowing only enough to get end-user things i want to use built for `conda-forge`. I learn a little bit each time, then forget it, then have to relearn when something major happens in the rust world. So it goes!

> The build script

Yeah, it looks like the newest `wasm-bindgen` would handle at least [one of the cases](https://github.com/rustwasm/wasm-bindgen/issues/2605). But multiple targets probably make a lot of sense, even if distributed in the same package.

**Comment by Tpt** [comment]
> The current effort has as a goal enabling jupyterlab-knowledge-graph. In a classic client-server setting (JupyterLab), pyoxigraph would provide a better out-of-the-box-plus-one (over rdflib-sqlalchemy on sqlite3) for a "real" jupyter server until sharing data between many users, megatriples, etc. are required and a distributed SPARQL store will make sense.

Ok! Thank you! I hope Oxigraph should work properly with a few million triples, at least for simple operations. Gigatriple is going to be an other beast. My goal is having Oxigraph work with at least a billion triples on commodity hardware but I am not there yet.

> so thank you for making things that would Just Work (Better) with existing code.

Thank you! I had rdflib code and I wanted to make it work with Oxigraph. I still have some trubles to use the SPARQL update Oxigraph implementation instead of the rdflib one but SPARQL queries should be natively executed by Oxigraph (oxigraph/oxrdflib#4).

> But it's rather a heavy opinion, so indeed might make the most sense for me to split up my prototype into several packages, one of which is an opinionated oxigraph wrapper. Once we can talk about it more directly, we can decide what can be upstreamed.

Thank you! It sounds like a great plan!

> As a heads-up: I will probably package both the rdflib wrapper and the underlying python bindings for conda-forge when i get a chance. I anticipate packaging it being much easier than some of the other stores!

Thank you! A plain "pip install" should work hopefully. I try to provide pre-build pyoxigraph wheels for the most commons plateforms. "oxrdflib" is just a plain python package.

>> The build script
>
> Yeah, it looks like the newest wasm-bindgen would handle at least one of the cases. But multiple targets probably make a lot of sense, even if distributed in the same package.

Yes, it sounds like a good plan. I have never built multiple targets node packages, help in this area will be much welcomed.

**Comment by mihai-dinculescu** [comment]
@bollwyvl can you please detail how you went about loading the wasm file with arraybuffer-loader?

My attempts oscillate between errors of `magic header not detected` and `failed to match magic number`.

**Comment by bollwyvl** [comment]
yeah, sorry: there are enough sharp edges between both of our unpublished code examples for me to be of much use... I will lobby for publication of at least my hacky example.

**Comment by mihai-dinculescu** [comment]
I've got it working eventually.

```js
// oxigraph.js
- const path = require('path').join(__dirname, 'oxigraph_bg.wasm');
- const bytes = require('fs').readFileSync(path);

- const wasmModule = new WebAssembly.Module(bytes);
- const wasmInstance = new WebAssembly.Instance(wasmModule, imports);
- wasm = wasmInstance.exports;
- module.exports.__wasm = wasm;

+ module.exports.wasmImports = imports;

+ module.exports.setWasm = function (instance) {
+     wasm = instance;
+ }
```

```js
// test.js
import oxigraph from 'oxigraph';

/// ...

const oxigraphBuffer = require('oxigraph/oxigraph_bg.wasm');
const bytes = new Uint8Array(oxigraphBuffer);
const wasm = await WebAssembly.compile(bytes);
const instance = await WebAssembly.instantiate(wasm, oxigraph.wasmImports);
oxigraph.setWasm(instance.exports);
```

```js
// webpack.config.js

{
    test: /\.wasm$/,
    type: 'javascript/auto',
    use: [
        {
            loader: 'arraybuffer-loader',
        },
    ],
},
```

**Comment by bollwyvl** [comment]
Nice!

I think i also had some webpack plugin handle the patching, and went down the comlink route, but that's a great starting point!

**Comment by mihai-dinculescu** [comment]
Yes, I too went the string-replace-loader path. Here's the corresponding config for the above.
Thanks for your initial post. It was a great starting point.

```js
// webpack.config.js
// ...
{
    loader: 'string-replace-loader',
    options: {
        multiple: [
            {
                search: 'const { TextDecoder, TextEncoder } = require(String.raw`util`);',
                replace: '',
            },
            {
                search: "const path = require('path').join(__dirname, 'oxigraph_bg.wasm');",
                replace: '',
            },
            { search: "const bytes = require('fs').readFileSync(path);", replace: '' },
            { search: 'const wasmModule = new WebAssembly.Module(bytes);', replace: '' },
            {
                search: 'const wasmInstance = new WebAssembly.Instance(wasmModule, imports);',
                replace: '',
            },
            { search: 'const wasmModule = new WebAssembly.Module(bytes);', replace: '' },
            { search: 'wasm = wasmInstance.exports;', replace: '' },
            {
                search: 'module.exports.__wasm = wasm;',
                replace: `
module.exports.wasmImports = imports;

module.exports.setWasm = function (instance) {
wasm = instance.exports;
}
            `,
            },
        ],
    },
},
```

And the optional Decoder replacement which I haven't tested yet.
```js
// webpack.config.js
// ...
    plugins: [
        new webpack.ProvidePlugin({
            TextDecoder: ['text-encoding', 'TextDecoder'],
            TextEncoder: ['text-encoding', 'TextEncoder'],
        }),
    ],
```

**Comment by Tpt** [comment]
> Would this be a way to minimize the JS/Rust/WASM conversion cost issue? i.e. load a bunch of triples in a single transaction, passed into WASM as a turtle string?

Yes!
